### ISSUE TITLE
[RFC-1.17] The Fluid Size of a Bucket

### ISSUE THREAD

**Issue by Shadows-of-Fire** [MinecraftForge/MinecraftForge#8166]
This exists to collect feedback about possibly changing the volume of the bucket from 1000 to a number that is more divisible with commonly utilized sub-increments.    
Currently, the volume of a bucket is defined as `public static final int BUCKET_VOLUME = 1000;`  
This has issues even when moving down to the next most commonly utilized fluid increment, 3 (for bottles).
Below is a table of possible bucket values and common divisors that are used, alongside the quotients.  

![](https://i.imgur.com/ppMe1Xr.png)

The important thing to note here is how many quotients are whole numbers, compared to the raw size of the bucket value (as lower numbers are better, as to allow for the continued use of integers).
  
The two most promising candidates are **1296** and **5184**.  The "perfect" candidate, at least for all of the known divisors, is **129600**, but this value would practically make the continued use of integers unreasonable.

I don't think swapping to longs is a great option for fluidstacks, but I'm not sure if there is a significant overhead to it, so it may be an option.

**Comment by Alpvax** [comment]
I would suggest using something like 1200, because it is easily divisible by 3,4 and 10. The only question is whether the lack of a nice number for /9 is a deal-breaker.
My vote would be for 1296 if not, because I feel that moving up to 5k+ is a bit excessive.

**Comment by KnightMiner** [comment]
Tinkers Construct has always wanted something divisible by both 4 and 81 for our purposes. Would let us make a bucket of fluid be the same amount as a block of metal. No matter what number is chosen, we will keep using values that ensure those factors, but it would be nice to integrate the cauldron and glass bottles with our fluid values without worry of duplication bugs.

**Comment by gigaherz** [comment]
1200 would be a really good number (16x75, 100x12), if it wasn't commonplace to have melted metals.

1200 not being divisible by 9 means it's not able to represent 1 bucket == 1 block == 9 ingots. we could say that 9 ingots are not equal to a bucket, but it complicates things. And since it doesn't represent even whole ingots, it also doesn't represent nuggets, which would need 81 as a divisor.

Edit: was typing on the phone and didn't notice KMs reply.

**Comment by Alpvax** [comment]
That is a good point, which is why I concede that 1296 would be better.

My biggest worry is that people won't realise that the bucket quantity has changed, leading to dupe bugs.
Also, using 1296 reeks of an imperial system (although it makes dividing it up easier). We would have to make sure we move away from the "mili"-bucket prefix.

**Comment by Lykrast** [comment]
The /81 really makes things a lot uglier, but I see the use for it with nuggets (even if TiCon is the only mod I know of that needs it). Not having /5 and /10 and having the numbers be so "complicated" (imo 125/1000 is much easier to grasp than 162/1296) is def an issue with several tech mods that use fluids.

However, with something like 1296 there's still /4, /8, /12, and /16, and since vanilla operates more on /8 (like coal burn time), /9 (metal compression), /16 (snowball stacks and stuff), /64 (normal stacks) I feel like they would "fit more" than the /5 and /10 that some mods used for fluid crafting (even if they were easy to grasp). It'll just be a matter of rebalancing the resource costs in the recipes to match the "new" ratios (if they weren't already on /4 /8 and such).

**If I had to vote I'd do for 1296**, since it's the closest to a round 1000 and also has nice divisors. I don't think having /64 for the 5184 is worth the int size problems and the weirdness that we're diving a bucket by ~5000 instead of ~1000. No matter the value, inertia is gonna be a problem anyway so I'm hoping the Big Mods will transition smoothly.

And also we definitely need a new name for the mB since the mili- is no longer appropriate. And *maybe* something nice to help display fluid values in `X buckets + Y units` since it's gonna be much harder to tell at a glance how much buckets does a lot of units make, while it was very easy with the 1000.

**Comment by KingLemming** [comment]
Oh Christ, not this again.

**Comment by hanetzer** [comment]
Imperial buckets wen? Also voting 1296. To elaborate, in vanilla minecraft *nothing* is metric that impacts gameplay, and the only metric thing at all is purely arbitrary (the size of a block being a meter square). It could have just as well been yards or just 'blocks' in the distance traveled statistic.

**Comment by Alpvax** [comment]
To be completely honest, although I can see the argument, I think it is actually better (for the players) to stick with the metric bucket. Mods like TiCon could use 999, and just say that 1mb is lost melting it down. In other words, 1 block of metal is 999mb.

Alternatively, mods could use fractions of (milli)buckets and just describe them as e.g. "3 nuggets worth of molten iron".
However, the fact that bottles equate to 1/3 of a bucket does sort of throw a spanner in the works.
If it wasn't player visible, I would say we should switch to 1296.

As it is, I think it might be worth keeping 1000 as a factor and  increase the number to 81,000 which seems excessive. However, then we can continue to display amounts to players in mb, but behind the scenes we can track it in smaller increments, divisible by 2, 3, 4, 5, 8, 9 and 10.
This would mean that you can choose to display in mb (81 units per mb) or bottles/ingots/nuggets worth.

**Comment by PrincessOfEvil** [comment]
I'm not a big fan of actively lying to the player. At that point, might as
well show the value in buckets and fractions, the way any machinery that
works with molten metals does it at the moment.

Voting for 1296, as the bigger ones noticeably eat into the int. Though
1296000 with longs is not a terrible idea.

**Comment by voidsong-dragonfly** [comment]
Going to pop in and note about the 81000 or larger things - I remember some talk on the Fabric side from players who were very very annoyed about having to use one of the indev? something? fluid APIs that had 81000, general consensus I remember was that 81000 is much too large to be useful. 

Re: Bottles could be dealt with by making them default to 250mB if 1000 was stuck with, there are already multiple? at least one large mod(s) that do this by default and while it does break vanilla default slightly, I feel like it is a very small change.

Final thing for me is that since 1000mB to a bucket also means 1mB is one liter, changing it from 1000 would break the 1mB == 1L symmetry, which I would heavily dislike.

**Comment by Alpvax** [comment]
> it does break vanilla default slightly, I feel like it is a very small change.

Forge is designed such that with no mods, it functions exactly like vanilla, so this would never be accepted.

>  81000 is much too large to be useful.

This is the only reason I'm not keen on it, otherwise it would be my first choice

> Bottles could be dealt with by making them default to 250mB if 1000 was stuck with

That still wouldn't fix the 81 nuggets -> 9 ingots -> 1 block issue, which is one of the biggest reasons to change.

I also like the fact that buckets are currently metric, which is what the majority of the world uses.

**Comment by TeamSpen210** [comment]
Perhaps something to alleviate the change would be to have Forge also include code to format the internal number into fractional values - buckets, bottles, etc. Nothing in MC is actually metric, so trying to stick to that doesn't really fit well. If showing fractions, the actual number would just be something used internally.

**Comment by voidsong-dragonfly** [comment]
While nothing in MC is inherently metric, I would argue the tech mods using the fluid system all use some form of the metric system for simplicity so sticking it to it is incredibly helpful for them.

**Comment by KnightMiner** [comment]
Honestly, most tech mods I have seen either deal with buckets or make up some unit. E.g. thermal expansion has 100mb per redstone (900mb per block), or 250mb per potion (disagreeing with the cauldron and making it 1/4). Immersive Engineering also does 125mb for treated wood (1/8) I don't see an argument here that "tech mods" want metric, they seem to want factors of 4, 8, and 9 same as other mods.

You could probably give an example of a mod that uses metric fluid units, but you cannot say all tech mods do so when two of the biggest ones don't make use of factors of 10.

**Comment by ShetiPhian** [comment]
This keeps being brought up, but nothing gets agreed upon because no number covers all cases.

These are a given:
1 (Bucket / Block), 3 (Bottle), 9 (Ingot), 81 (Nugget)

Are other numbers truly needed?
Shouldn't care about common stack sizes as this is fractions of a single.
This isn't metric so it doesn't have to be divisible by 10.
If need be, anything currently using half or quarter can change to thirds thus matching bottles.
Are sixteenths, sixty-fourths, and one-hundredths ever used?

The suggested 1296 fits well, covers the must haves, and the half/quarter that is used now.
1296 - (2 | 648), (3 | 432), (4 | 324), (6 | 216), (8 | 162), (9 | 144), (12 | 108), (16 | 81), (18 | 72), (24 | 54), (27 | 48), (36 | 36)
It would be my primary pick.

------------------------------------
Has lower values been looked into? I've only seen above 1000. 
(Lower also allows more buckets to fit in an integer)

For example:
8100 was mentioned but why not use 810?

810 makes bottles 270, Ingots 90, Nuggets 10
While still enabling half (405), fifth (162), and tenth (81)

quarter, and eighth fail though so they would need to be changed to to the closest lower or higher value
third(270) / fifth(162) & sixth(135) / ninth(90) 

810 - (2 | 405), (3 | 270), (5 | 162), (6 | 135), (9 | 90), (10 | 81), (15 | 54), (18 | 45), (27 | 30)

**Comment by Shadows-of-Fire** [comment]
If the values of 16 and 64 are not really worth considering (which may very well be the case, I don't know of any currently held or desirable uses to represent 1/16th or 1/64th of a fluid), then the number **324** works very well.

Updated Table:
![](https://i.imgur.com/IARLpC1.png)

**Comment by PrincessOfEvil** [comment]
Glass panes and metal bars are 3/8ths, but i don't recall any uses for sixteenths, so 648 might be a worthy consideration as well.

**Comment by Alpvax** [comment]
> These are a given: 1 (Bucket / Block), 3 (Bottle), 9 (Ingot), 81 (Nugget)
> 
> Are other numbers truly needed?

The biggest advantage of buckets being metric is that players can tell exactly how many buckets are in a tank at a glance.

**Comment by TheSilkMiner** [comment]
It is not at all required to use a metric system for that. The Mod responsible for showing how much fluid is in a tank can simply operate the conversion. Instead of showing `2600 mB` like it does now, it can show `2 buckets and 8 fluid units` or something similar. For this reason, I'd argue that the way stuff is shown to the player is irrelevant.

**Comment by voidsong-dragonfly** [comment]
Yes and no, showing units in the imperial style ('N units X and M units Y and A units Z') can be both tiring and annoying unlike dealing with a metric-like system of 1000 mB == 1B and not
'I have to get 3 and 1/9 buckets for this to work, and I have 2 1/8 here and half a bucket there, how much more do I need to produce'
With mB that's something like 'I need to get 3250mB here, and I have 2125 mB and 500 mB, so I how much more? 625'
The tinkers' smeltery already does that, and it's a pain to deal with because adding mB, nuggets, blah blah blah gets tedious very quickly

**Comment by ShinyPorygon** [comment]
I'm going to shove this here from the CoFH discord, not my idea don't shoot the messenger. What if we just ditch fluid values all together and just go off labels "1 nugget of molten iron" "one water bottle of water" it loses some flexibility but there would be no numbers to debate about.

**Comment by JaisDK** [comment]
What exactly is being achieved by changing the bucket size to a different value that _also_ doesn't work in all scenarios?

I'm gonna throw my vote behind the above suggestion. Have different fluid types depending on the item you're converting to fluid and label them accordingly. Metals would be labelled with block/ingot/nugget, gems as gem, potions and water bottles would be labelled as "2 bottles of x", Glowstone, Redstone and other materials would just use the name as label "1 Redstone".
It brings other challenges, like how to define the size of a fluid container, but at least you don't have to worry about the equivalence of iron ingots with water bottles.

**Comment by ShetiPhian** [comment]
A number is required for conversion and storage.

What the user is shown does not matter at this point.
Displaying the value is not limited to whole numbers, like storage is.
2600 fluid can simply be displayed as 4.01 buckets (based off 648 being a bucket)
3250 would be 5.02

**Comment by ShinyPorygon** [comment]
I'd like to throw in my actual opinion on this. Honestly just don't mess with it. By now most mods have their way of doing it with the current system. If you change that then the whole balance will be disrupted. 

Obviously not everything is 100% perfect as is but if it ain't broke don't fix it. Instead try and make more specific values known. Ingot = 1296mb, Bottle = 250mb/333mb.

**Comment by gigaherz** [comment]
> What if we just ditch fluid values all together and just go off labels

That makes it impossible to have any sense of volume when dealing with general-purpose pipes and tanks.

How full does "1" make a tank, if it's fluid X? 
How many seconds should it take to move 1 Y through a tier 1 pipe?

...

I'm starting to get the thought that maybe we are looking at this wrong: Just like different items have different max stack sizes, we could make Fluid attributes include the bucket volume for that fluid, and they could also have a method to display the contained amount. Forge could then default to a good number like 1000 (displayed as mB), for anyone who doesn't have a hard requirement.

There's also a big downside to every single conversation we keep having about this: it gives no standard measure for what other volumes are. Even if we allowed different mods to have different volume fractions, we can still have one mod may say "1 block = 1 bucket, so our molten block is 81, our molten ingot is 9, and our molten nuggets is 1 unit exactly", while another may say "we consider a block to be 1.25 buckets, so we want our bucket volume to contain 243 subdivisions, where 1 nugget = 4, 1 ingot = 36, and 1 block = 324". Is that something we also need to fix? or can we at least agree on 1 bucket = 1 molten block = 81 molten nuggets?

Edit:

>  Ingot = 1296mb, Bottle = 250mb/333mb.

1 Bottle = 333mB means when you take out 3 bottles, you have 1mB left over in the bucket/tank. Those rounding errors are the **entire reason** this conversation happens. That's what is being fixed

**Comment by JaisDK** [comment]
You would quickly run into duplication issues if you have two fluids that are tagged as the same but have different bucket volumes.

**Comment by gigaherz** [comment]
How so exactly? Please try to reason through arguments not just throw concerns around without explanation.

**Comment by JaisDK** [comment]
Sure, if you have a "filler" container that fills buckets with fluids and which can also be filled with fluids from buckets. You can then fill up the "smaller" bucket sizes, convert the "small" bucket to the other type and fill up the container using that "big" bucket.

**Comment by JaisDK** [comment]
You would need to be able to convert between the two buckets, either directly or through crafting.

**Comment by dudebehinddude** [comment]
I vote 1296, and display to the user like '2.33 buckets (2 blocks, 3 ingots)' and show millibuckets/whatever the unit is changed to if advanced tooltips are enabled

**Comment by KnightMiner** [comment]
> You would quickly run into duplication issues if you have two fluids that are tagged as the same but have different bucket volumes.

You would quickly run into duplication issues if you have two items that are tagged the same but have different costs. You would quickly run into duplication issues if you call two things the same that are not the same. I don't get your point, if they have different bucket volumes/fluid rates, don't tag them the same

----

Edit:

> I'm starting to get the thought that maybe we are looking at this wrong: Just like different items have different max stack sizes, we could make Fluid attributes include the bucket volume for that fluid, and they could also have a method to display the contained amount. 

The biggest problem is with stack sizes, "1" is the minimum unit. With fluids, "1" contains a large amount of units. This would make fluid pipes more complicated if its done. Instead of saying "my pipe transfers 100mb per second", I now have to say something like "my pipe transfers 1/10 of a bucket" and if its not divisible by 10, I have no choice but to transfer amounts that won't fill a tank to even bucket amounts. This is something we can work around, but its important to note that is the result of such a decision.

Id also ask what sort of storage system we would use for that. Are fluid stacks still storing an int, and we have to perform bucket math every time we want to check tank capacity? It would feel wrong if your 10000 unit per bucket fluid cannot fit in my 1 bucket tank because I made its size 1000 fluid units instead of considering the fluid's bucket value

**Comment by gigaherz** [comment]
> Sure, if you have a "filler" container that fills buckets with fluids and which can also be filled with fluids from buckets. You can then fill up the "smaller" bucket sizes, convert the "small" bucket to the other type and fill up the container using that "big" bucket.

I think you are misunderstanding, which is my own fault. The idea was to make the units fluid-dependant, not the actual size of the bucket! 1 bucket of X and 1 bucket of Y are the same actual volume, just have different fractions in it. A bucket is always the same physical dimension: however much space is needed to contain 1 whole block of water.

I should have used a different name for the method, `getBucketDivisions` or such, I only used `getBucketVolume()` because we currently have a constant called `BUCKET_VOLUME` which is set to 1000.

>  It would feel wrong if your 10000 unit per bucket fluid cannot fit in my 1 bucket tank because I made its size 1000 fluid units instead of considering the fluid's bucket value

Well, yes. The idea would mean you don't store units, you store "buckets". A tank should have space for say, 20 buckets, or 0.5 buckets. If the FluidStack is not divisible in a good way you just round down, but **always** store using the fluidstack chosen divisor, not some arbitrary one. With that idea **everything** is expected to respect the fluidstack's chosen divisor.

EDIT: To clarify, I don't know if the idea I proposed is actually good, and I'm just giving it as a modder, not as a member of the forge team.

**Comment by tterrag1098** [comment]
Because this come up every so often, and always goes the same way, I will just repost the same response I've used the last 5 times it came up.

TL;DR: Blocks, Ingots, and Nuggets are not units of fluid, and shouldn't be a concern for this decision. Their fluid-equivalent nature is the decision of ONE mod.

> basing your entire API on the ideas of one mod is really shortsighted
> If bottles are the big issue then the simple solution is to just make them your root unit
> 1 bottle = 1 bottle
> for gameplay, you add some zeros for smaller amounts, call those "drops" or whatever
> 100(0) drops = 1 bottle
> 3 bottles = 1 bucket
> API complete
> why does ingots and nuggets need to come into it? Those are strictly NOT fluid shapes
> Why do they need to be related to water volume at all?

**Comment by PrincessOfEvil** [comment]
TT, pls. Even bottles are simultaneously three and four in a bucket.

The whole problem isn't ONE mod (and it isn't even _one_ mod, i know of at least four, and that's an entire discussion in of itself), it's that 10 is not a number Minecraft makes use of. All crafting, smelting and other operations work off powers of two or three. 

Making metals rounded to a power of 10 is very silly, since workbench power of 9 conversion is lossless and thus any other lossless conversion will create duplication issues.

Transferring 1/10th of a bucket is a fair consideration in a system with a variable bucket, but it's a little funny with buckets set in stone, since most numbers in tech mods are rather arbitrary.

Also, fun fact: the metric system allows use of units that aren't scaling by powers of ten, if having a different base is better. Angles (base 60, and yes, you are allowed to use degrees, it's not illegal within SI) or data (base 2) come to mind.

**Comment by Mishura4** [comment]
1296 makes the most sense. It's close to 1000, it's divisible by 3, 9, 81 which fixes the metals and bottles issues, it's divisible by 2, 4, 8, 16 which allows for precise fractions and values and also allows for integration with mods that divide by 16 like Chisel & Bits if ever needed, as well as snowballs.

"If it ain't broke don't fix it" is just sabotaging yourself really. We change things not because they're broken but because they can be and we can make them better. We get and enjoy Minecraft updates not because the game is broken but because new features with new gameplay possibilities are added. But also it *is* broke, that's the point, because bottles in vanilla are 1/3 bucket, right now you either have to break vanilla behaviour which is never a good thing in a mod, or you have to add weird spaghetti code like adding a special case to remove 1mb when converting 3 bottles, how do you even keep track of the 3 bottles when you add 2 bottles then 1 bucket then 3 ingots then come back later and add 1 bottle, how about 2 bottles then transferring them to a container then processing them then adding 1 bottle, etc you just can't.

Also the metric system is never used in Minecraft. You don't count how many ingots you have in a chest in kiloingots or centiingots or whatever, you count them in stacks. How many iron ingots do I have in my chests? I have 3 stacks plus 37 ingots plus 3 nuggets, or 229 (if i calculated this correctly, i suck at mental math) ingots and 3 nuggets, not 229.3333 ingots or 0.2293333 kiloingots. At the end of the day I don't think that you "dislike" having to see "3 buckets + 1 bottle" in stead of "3333mb" in WAILA or TOP is worth giving up on the net positives of this change, you'll adapt and then you'll forget about how it was even different before. *And WAILA and TOP aren't meant to be a core feature of Minecraft anyways*.

The whole 1 bucket = 1000mb and ingots are 144mb etc were decisions made separately without much thought put into it, and then people kept using them for compatibility, but now what we've been using them for a while and can see their limitations, we can change that, and who else to do that better than the framework we already all use that's been providing standards for compatibility between mods for a while now already (e.g. Forge Energy). Even in vanilla without considering ingots 1 bucket = 1000mb breaks the bottle behavior, which is already enough of a reason to change it. No reason we can't incorporate metal compatibility in it at the same. It's less about the "shape" of a fluid as it is providing a unit that gives the most flexibility and the most modding possibilities, as is the purpose of Forge.

**Comment by Mishura4** [comment]
With that said I think gigaherz's idea of having the unit representing the amount contained in a bucket be provided by the fluid is a good alternative idea, one I can see is that yes pipes would have to rely on fractions of the amount, by which the unit provided might not be divisible. I don't think it would matter that much as you generally want to transfer until your tank is empty anyways, which would still be achieved eventually, even if the last tick doesn't transfer the regular amount. This would make it even more complicated for mods that display amounts however, but I would argue that's on them, maybe. Compatibility between mods might suffer a bit, if you want to make a mod over a mod providing an ingot for a fluid the original mod provides and you can't divide their unit by 9 for example.

**Comment by Alpvax** [comment]
> Also the metric system is never used in Minecraft.

Except that it has been used for the last decade for fluids. As far as I can recall, the first mod which added denominations to fluids was buildcraft (apparently version 2.1.0 which I believe appeared either late 2011 or early 2012).
If we change the amount of "drops" in a bucket, that is fine (I have no objection to changing it, and there are many good arguments to do so), but we **NEED** to make sure that it is obvious to players that the old numbers don't apply.

I am in favour of changing to a more divisible system, but not at the cost of making things more confusing for players.
gigahertz' system sounds intriguing, but we would need to move away from listing quantities in terms of "milibuckets", and I'm not 100% sure how it would interact with piping systems/tanks. What happens if a tank can store 1.5 buckets of fluid, but the fluid is not divisible by 2?
One potential solution I could see to this is to let the fluid dicatate the transfer speed as well (i.e. this fluid travels at 2 units/tick), then the piping systems could adjust that speed by either transfering every N ticks (for slower transfer) or transferring n*amount / tick (for faster transfer). However then we would need to decide on a sensible base speed for vanilla fluids (water & lava).
Then we would need to handle displaying and storing the quantities in third party tanks, which would be more complex.

**Comment by tterrag1098** [comment]
> TT, pls. Even bottles are simultaneously three and four in a bucket.

This is just a consequence of the current system. Nothing in vanilla implies that a bottle is 1/4 of a bucket.

> The whole problem isn't ONE mod (and it isn't even one mod, i know of at least four, and that's an entire discussion in of itself), it's that 10 is not a number Minecraft makes use of. All crafting, smelting and other operations work off powers of two or three.

The point wasn't how many mods have this kind of feature, but that each solves it slightly differently. None should be the driving reason behind a change in forge.

The necessity for fluids to be divisible by 9 is merely an assumption based on some current implementations of certain features. There is no reason for 1 block of metal to be equivalent to 1 bucket of liquid besides convenience. Just choose a new base unit (nugget) and use that. It's not something forge needs to be concerned about.

**Comment by Alpvax** [comment]
> There is no reason for 1 block of metal to be equivalent to 1 bucket of liquid besides convenience. Just choose a new base unit (nugget) and use that. It's not something forge needs to be concerned about.

Expanding on this, I remember having this discussion before, and pointing out that a bucket of fluid is actually a smaller quantity than a block (as in it fits inside a 1x1x1 block hole with a slight gap at the top). There is nothing stopping the metal mods using 1296 as the quantity of a metal block, and just accepting that it doesn't fit in a bucket.

The biggest argument I can see for changing the bucket quantity is that vanilla uses 3 bottles to 1 bucket, and currently the quantity of a bottle cannot be represented in mb

**Comment by TheCurle** [comment]
A bucket is simultaneously a full block (ie. when under another) and less than a full block (when on the surface)

This is something you can't really apply logic to.

**Comment by tterrag1098** [comment]
> The biggest argument I can see for changing the bucket quantity is that vanilla uses 3 bottles to 1 bucket, and currently the quantity of a bottle cannot be represented in mb

Which is exactly the problem my original comment addresses.

> This is something you can't really apply logic to.

Yep, which is why I suggest ignoring it entirely 😛

**Comment by Shnupbups** [comment]
_**Posting on behalf of someone else who would like to remain anonymous**_

### The choice of unit
Honestly 81000 is good enough for most purposes. The unit to use for fabric's fluid API was debated for more than 2 years. As soon as [the full API](https://github.com/FabricMC/fabric/tree/1.17/fabric-transfer-api-v1) (of note: very elegant imo, has transaction support, and is unified between items and fluids) was merged, discussion of the unit immediately stopped anyway. Just pick something that is reasonable and brings an actual improvement.

### long vs int
I haven't seen any viable reason not to use long, period. Any "memory usage" argument is rendered ridiculous when you consider the memory footprint of capabilities on every single ItemStack.

### What everyone forgets: cauldrons are not block entities
Is changing the denominator to a multiple 3 enough to support cauldrons? Not at all. That would requiring separating the capability system from block entities, which is not trivial at all with LazyOptional invalidation. Or you could drop LazyOptional<T> in favor of @Nullable T, which has been suggested multiple times already. For this too, it might be useful to [look at how Fabric addresses this](https://github.com/FabricMC/fabric/tree/1.17/fabric-api-lookup-api-v1).

**Comment by Mishura4** [comment]
The main argument against using long in favor of int is not memory size, but performance. 

https://stackoverflow.com/questions/19844048/why-is-long-slower-than-int-in-x64-java
(With that said this was on Java 7, also on Java 8, I don't know how it is on Java 16)

TL;DR Java optimizes int a lot, not so much long, sometimes splits long into two ints, even on a x64 JVM, and so you get slower execution to the point of being 10 times or more slower.

I don't really see any point of having extremely large values of fluids that would overflow an int anyways, if you need to store that much of an amount, why not use a second FluidStack similarly to how long is handled on 32 bit architecture with 2 ints? It seems such an extreme case that I'm not sure is even worth strongly considering sacrificing performance for it ; considering mods do operations on FluidStacks every or almost every tick, I think it's fair to be concerned about keeping int for it, and try to alleviate the overflow issue by picking a smaller unit.

**Comment by KingLemming** [comment]
> 
> _**Posting on behalf of someone else who would like to remain anonymous**_

Nope. There's no sense in that, sorry. Worst case, they can make a smurf account and post for themselves.

As far as the root of the discussion, @tterrag1098 already cut to the heart of it - vanilla *itself* is inconsistent (honey is 4, water is 3), and people are way too obsessed with trying to make some sort of nugget:ingot fluid ratio equivalency.

It's not necessary, and there are already points where ratios in vanilla completely break down (coal vs block of coal burn time, the water/honey disparity. In 1.17, LAVA in a cauldron...etc.)

Is 1000 perfect? No. Is it good enough? Yep, sure is. Can we close this? I think we can.

**Comment by voidsong-dragonfly** [comment]
> Is 1000 perfect? No. Is it good enough? Yep, sure is. Can we close this? I think we can.

Hard agree here.

**Comment by Mishura4** [comment]
You have an opportunity to make it better. Why wouldn't you 😆

**Comment by voidsong-dragonfly** [comment]
Because, as King said, it's already as good as you're going to get?

Vanilla is inconsistent so the "nuggets/ingots must have a block equivalent that is a bucket exactly" doesn't hold as a reason to change bucket size; the number of bottles in a block is inconsistent, (4 vs 3 for honey vs water), etc, etc. Buckets don't need to be exactly a block of metal so the main argument for changing it from 1000 goes away. Outside of block/ingot/nugget math, 1000 is a _very very_ good number to go with - it's round, divisible by a lot of numbers, a power of 10, 1mB == 1L, it's been 1000mB to a bucket for a long long time, etc. 

Without the block-bucket consistency argument, an argument that holds little to no weight due to MC itself being inconsistent, a number other than 1000 is not going to be a better number.

**Comment by KingLemming** [comment]
> 
> 
> You have an opportunity to make it better. Why wouldn't you 😆

Because the first rule of engineering is to not make things worse. And that's what this would do. To actually make it better, vanilla would be re-engineered to be consistent. Since that isn't being discussed here (or feasible), changing the number to an unintuitive value is actively detrimental.

The fluid system as it stands has some issues, but the amount of fluid stored in a bucket isn't one of them.

**Comment by Eelt** [comment]
As a mod developer that develops a mod that uses fluids, but not in explicit relation to blocks, ingots, nuggets, etc. this change would be fairly arbitrary for my mod and would just create an unnecessary hassle for me to rebalance my mod for 1296, 5184 or whatever other value would be used for the bucket volume. 

For example, my recipes that today under the 1,000 system that consumes 250 mB (1/4 of a bucket), would now become 324 units under 1296, and 1296 under 5184. Ironically, I think these numbers are pretty messy for fractional numbers of a bucket when contrasted with the 1,000 system we have today, and for me, which I don't base my fluids off of Blocks, Ingots, and nuggets, I would argue that is simply overcomplicating fluid volumes. Additionally, most players outside of America would find 1/4 of a bucket being 250 fairly intuitive as that is what most humans would more or less expect a quarter to look like being familiar with the metric system. The 9 system enforced by Minecraft's crafting table is not mathematically intuitive to players as it's not derived from a system that normal people use in their lives outside of Minecraft.

From a more technical standpoint, cutting down numbers say 1,000 to 1.0K would no longer be feasible in UI elements if moving away from the 1,000 system as I'd be forced to show every single-digit or deal with the horrors of writing more boilerplate to split my values into representations of blocks, ingots, and nuggets similar to how in the imperial system heights are given in feet, and then the remainder is converted in inches (except worse as there's basically 2 remainders now). And even then, splitting out fluids represented as blocks, ingots, and nuggets at the same time (eg 3 blocks, 2 ingots, 5 nuggets) wastes even more UI space, and for a mod developer working on mods that don't relate fluids in anyways to ingots or nuggets just makes any change away from 1,000 to essentially be a downgrade and give mod developers more busy work and potentially cause confusion or frustration for the players.

The one thing I think proponents of this change have missed is what do players and the border community think about this? How would they react to this? Do people who play our mods or use Forge really want to dedecimalize our fluids into the mess of writing out units as blocks, ingots, nuggets and have to worry about the additional, unintuitive, calculations of conversion between these units, especially when they make no sense (mods that don't moltenize metals, that don't relate fluids to ingots, and nuggets explicitly, etc). I think that this level of change needs to address the Modded Minecraft community as a whole rather than just a select few mod/Forge developers who follow Forge RFCs. 

At the end of the day, I'm not convinced that 1296 or 5184 for the global bucket volume value would make _players_ lives easier, and that's what should matter the most.

**Comment by LexManos** [comment]
vivat milli situla

**Comment by Shnupbups** [comment]
Posting as myself now, but how is 1000 "good enough"? Literally the *only* standards for fluids in vanilla are Bottles and Buckets, and with Water Bottles being 1/3 of a Bucket that is immediately something that can't work with 1000.

When your denominator is incompatible with the *main example* of vanilla fluids, it's bad.

**Comment by Alpvax** [comment]
> When your denominator is incompatible with the _main example_ of vanilla fluids, it's bad.

It is, but the only solution without the massive metric trade-off is to make bottles the base unit. But there are 2 conversions for bottles to buckets, so that doesn't work.

There is no good solution without losing the metric system (and all player knowledge built up over the past decade), and no consistency in vanilla to base a better system on.

At that point, the biggest factor in the decision is consistency, and keeping it at the 1000 that it has always been has more advantages than disadvantages IMO.

**Comment by KingLemming** [comment]
> 
> 
> Posting as myself now, but how is 1000 "good enough"? Literally the _only_ standards for fluids in vanilla are Bottles and Buckets, and with Water Bottles being 1/3 of a Bucket that is immediately something that can't work with 1000.
> 
> When your denominator is incompatible with the _main example_ of vanilla fluids, it's bad.

Most mods have already settled on using 250 mB for a bottle, which works just fine, because water is trivially renewable so nobody actually has to care. It also goes neatly with fluid XP amounts and a bottle of enchanting being approximately 12.5 XP (averaged).

With honey, vanilla has also established 250 mB / bottle, since it's 4 bottles to make a block. And going forward, you can put Lava in a cauldron, but not in a bottle. So...what's the equivalency there? There isn't one.

Vanilla is inconsistent. The notion that the entire system needs thrown out and made unintuitive to accommodate a poor design choice by vanilla MC is nonsense.

**Comment by Technici4n** [comment]
I feel like 81000 retains all the advantages of 1000, especially when displayed as decimal millibuckets, but on top of that allows nuggets to be exactly 1/81th of a bucket.

And yes nobody cares about water cauldrons or water in general, but it sucks that you can't extract lava from lava cauldrons using pipes, as that is something that I'm sure some players would like to be able to do.

(On a side note, it also sucks that you can't insert items in composters, but that's not directly related.)
